[bugfix] Update of schema to work with singer target target-postgres

CStejmar opened the pull request:
The schema previously didn't include any of the currencies, which resulted
in that only the date property was added to the database. The records
include correct currency data but they were simply not added when target-postgres
was used. Targets like target-csv and target-gsheet worked fine though.

Commit:
[bugfix] Update of schema to work with singer target target-postgres

The schema previously didn't include any of the currencies, which resulted
in that only the date property was added to the database. The records
include correct currency data but they were simply not added when target-postgres
was used. Targets like target-csv and target-gsheet worked fine though.

cmerrick commented:
Hi @CStejmar, thanks for your contribution!

In order for us to evaluate and accept your PR, we ask that you [sign a contribution license agreement](https://stitch-cla-enforcer.herokuapp.com/). It's all electronic and will take just minutes.

cmerrick commented:
You did it @CStejmar!

Thank you for signing the Singer Contribution License Agreement.

awm33 commented:
There's probably a programmatic way to do this as well, eg during the discovery hit the API and get the list of currency codes. There are dynamic schemas in other taps.

If the list of currencies doesn't really change (whenever a new country/currency is formed?), then maybe static is just fine.

CStejmar commented:
@awm33 Yes, there's probably a dynamic way of doing it. I can take a look at a programmatic solution instead. As new currencies are seldom introduced and old ones seldom removed I thought this static approach would be fine. But I checked the historical data from ECB and realised that that isn't the case. For example, Estonia introduced EUR in 2011 and abandoned EEK at the same time :). I will rework this and push a dynamic solution

Commit:
Dynamic update of currencies/rates in schema

If the currencies have changed for a date, the schema is updated
and rewritten. This will ensure compatibility with historic and
future data.

whatnick commented:
Can we get this merged please. Planning to use target-postgres.

CStejmar commented:
@awm33 Just pushed an updated of this PR that makes the solution dynamic

awm33 commented on tap_exchangeratesapi.py:
```diff
@@ -48,23 +49,31 @@ def request(url, params):
+                    schema['properties'][rate] = {'type': ['null', 'number']}
+
+            # Only write schema if it has changed
+            if schema != prev_schema:
```
@CStejmar Can schemas vary between days? If the API always returns the same currencies, or realistically within the same tap run, I would just emit the schema once on the first returned response.

CStejmar replied:
@awm33 Yes, it can vary between days. For example for EEK as I described above. I also noticed that ISK, for example, was added 2018-02-01. So if you would do a tap run from 2018-01-31 to 2018-02-01 you would not get the ISK into the database if you just emit the schema once. But yes, it does not change very often. What do you think, should I keep it as it is or change? I think it is more robust if we keep it as it is, but change `additionalProperties` to `False` as you suggested.

awm33 replied:
@CStejmar Ah, ok, then I think updating the schema each call makes sense.

CStejmar replied:
Ok, I will go ahead and change the `additionalProperties` 👍

awm33 reviewed (commented):
@CStejmar I would also set `additionalProperties` to `False` so it throws an error on schema validation.

Commit:
Changed additionalProperties to False in schema

timvisher commented:
This should be merged and released very soon. Thanks for doing the work!
